### PR TITLE
typescript: Fix enum emit when values are strings

### PR DESCRIPTION
**Pull request by andy-ms**
| Q                        | A
| ------------------------ | ---
| Fixed Issues?            | #7126, #7128
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR         |
| Any Dependency Changes?  | No
| License                  | MIT

When the value is a string, emit `E["B"] = "";` and not `E[E[""] = "B"] = ""`.
Also starts handling strings in the set of previously-seen values.
CC @DanielRosenwasser for review.

### CODE REVIEW

File: packages/babel-plugin-transform-typescript/src/enum.js [hunk 1]
```diff
@@ -88,21 +90,27 @@ function enumFill(path, t, id) {
...
  *   }
  */
-type PreviousEnumMembers = { [name: string]: number | typeof undefined };
+type PreviousEnumMembers = { [name: string]: number | string | undefined };
 
 function translateEnumValues(path, t) {
   const seen: PreviousEnumMembers = Object.create(null);
   // Start at -1 so the first enum member is its increment, 0.
-  let prev: number | typeof undefined = -1;
+  let prev: number | undefined = -1;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-typescript/src/enum.js]: `undefined` isn't  flow type. It should either be `typeof undefined` or `void`.

File: packages/babel-plugin-transform-typescript/src/enum.js [hunk 2]
```diff
@@ -88,21 +90,27 @@ function enumFill(path, t, id) {
  *     Z = X | Y,
  *   }
  */
-type PreviousEnumMembers = { [name: string]: number | typeof undefined };
+type PreviousEnumMembers = { [name: string]: number | string | undefined };
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-typescript/src/enum.js]: Can this actually be`undefined`? Looking at the `translateEnumValues` it seems to me that values are only set when they are a number.

File: packages/babel-plugin-transform-typescript/src/enum.js [hunk 3]
```diff
@@ -88,21 +90,27 @@ function enumFill(path, t, id) {
...
       const constValue = evaluate(initializer, seen);
       if (constValue !== undefined) {
-        value = t.numericLiteral(constValue);
-        prev = constValue;
+        seen[name] = constValue;
+        if (typeof constValue === "number") {
+          value = t.numericLiteral(constValue);
+          prev = constValue;
+        } else {
+          value = t.stringLiteral(constValue);
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-typescript/src/enum.js]: Can you add an assertion here? Something like
```js
else if (typeof constValue === "string") {
  value = t.stringLiteral(constValue);
  prev = undefined;
} else {
  throw path.buildCodeFrameError("Invalid enum value");
}
```

To disallow things like
```ts
enum Enum {
  baz = true
}
```

Comment (andy-ms) [reply packages/babel-plugin-transform-typescript/src/enum.js]: For `baz = true` we will simply not have a `constValue` -- it won't be set to anything other than a number or string. Still, added `assert(typeof constValue === "string");`.

**Comment by DanielRosenwasser** [comment]
Looks like the test needs to be run again, but otherwise seems good to go.

**Comment by existentialism** [comment]
@DanielRosenwasser restarted!

**Pull request merged**